fvm/src/syscalls/crypto.rs

```diff
@@ -4,46 +4,66 @@ use std::cmp;
 use fvm_shared::crypto::signature::{
-    SignatureType, SECP_PUB_LEN, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE,
+    BLS_PUB_LEN, BLS_SIG_LEN, SECP_PUB_LEN, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE,
 };
 use fvm_shared::piece::PieceInfo;
 use fvm_shared::sector::{
     AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
     WindowPoStVerifyInfo,
 };
 use fvm_shared::sys;
-use num_traits::FromPrimitive;
 
 use super::Context;
 use crate::kernel::{ClassifyResult, Result};
 use crate::{syscall_error, Kernel};
 
-/// Verifies that a signature is valid for an address and plaintext.
+/// Verifies that a bls aggregate signature is valid for a list of public keys and plaintexts.
 ///
 /// The return i32 indicates the status code of the verification:
 ///  - 0: verification ok.
 ///  - -1: verification failed.
-#[allow(clippy::too_many_arguments)]
-pub fn verify_signature(
+pub fn verify_bls_aggregate(
     context: Context<'_, impl Kernel>,
-    sig_type: u32,
+    num_signers: u32,
     sig_off: u32,
-    sig_len: u32,
-    addr_off: u32,
-    addr_len: u32,
-    plaintext_off: u32,
-    plaintext_len: u32,
+    pub_keys_off: u32,
+    plaintext_lens_off: u32,
+    mut plaintexts_off: u32,
 ) -> Result<i32> {
-    let sig_type = SignatureType::from_u32(sig_type)
-        .with_context(|| format!("unknown signature type {}", sig_type))
-        .or_illegal_argument()?;
-    let sig_bytes = context.memory.try_slice(sig_off, sig_len)?;
-    let addr = context.memory.read_address(addr_off, addr_len)?;
-    let plaintext = context.memory.try_slice(plaintext_off, plaintext_len)?;
+    // Check that the provided number of signatures aggregated does not cause `u32` overflow.
+    let pub_keys_len = num_signers
+        .checked_mul(BLS_PUB_LEN as u32)
+        .ok_or(syscall_error!(
+            IllegalArgument;
+            "number of signatures aggregated ({num_signers}) exceeds limit"
+        ))?;
+
+    let sig: &[u8; BLS_SIG_LEN] = context
+        .memory
+        .try_slice(sig_off, BLS_SIG_LEN as u32)?
+        .try_into()
+        .expect("bls signature bytes slice-to-array conversion should not fail");
+
+    let pub_keys: &[[u8; BLS_PUB_LEN]] = context.memory.try_chunks(pub_keys_off, pub_keys_len)?;
+
+    let plaintext_lens: &[u32] = context
+        .memory
+        .try_slice(plaintext_lens_off, num_signers * 4)
+        .map(|bytes| {
+            let ptr = bytes.as_ptr() as *const u32;
+            unsafe { std::slice::from_raw_parts(ptr, num_signers as usize) }
+        })?;
+
+    let mut plaintexts = Vec::<&[u8]>::with_capacity(num_signers as usize);
+    for len in plaintext_lens {
+        let plaintext = context.memory.try_slice(plaintexts_off, *len)?;
+        plaintexts.push(plaintext);
+        plaintexts_off += len;
+    }
 
     context
         .kernel
-        .verify_signature(sig_type, sig_bytes, &addr, plaintext)
+        .verify_bls_aggregate(sig, pub_keys, &plaintexts)
```

Comment:
So, we really try to avoid `O(n)` work before charging for gas. In this case, I'd pass a `plaintext_lens` and `plaintexts` (buffer) directly into `verify_bls_aggregate`. Otherwise, the user could force us to do a lot of work even if they don't actually have the gas to pay for it.

Specifically:

```rust
fn verify_bls_aggregate(sig: &[u8; BLS_SIG_LEN], pub_keys: &[[u8; BLS_PUB_LEN]], plaintext: &[u8], plaintext_lens: &[u32])
```

This _also_ means we don't need to re-sum the lengths when charging for gas.